cpp/src/experimental/louvain.cuh

```diff
@@ -1205,350 +1207,49 @@ class Louvain {
-      //
-      std::tie(new_src_v, new_dst_v, new_weight_v) =
-        combine_local_edges(new_src_v, new_dst_v, new_weight_v);
-    }
-
-    //
-    //  Now I have a COO of the new graph, distributed according to the
-    //  original clustering (eventually this likely fits on one GPU and
-    //  everything else is empty).
-    //
-    current_graph_ =
-      detail::create_graph<vertex_t,
-                           edge_t,
-                           weight_t,
-                           graph_t::is_adj_matrix_transposed,
-                           graph_t::is_multi_gpu>(handle_,
-                                                  new_src_v,
-                                                  new_dst_v,
-                                                  new_weight_v,
-                                                  num_clusters,
-                                                  experimental::graph_properties_t{true, true},
-                                                  current_graph_view_);
+    std::tie(current_graph_, numbering_map) =
+      coarsen_graph(handle_, current_graph_view_, dendrogram_->current_level_begin());
 
     current_graph_view_ = current_graph_->view();
 
-    src_indices_v_.resize(new_src_v.size());
-
     local_num_vertices_ = current_graph_view_.get_number_of_local_vertices();
     local_num_rows_     = current_graph_view_.get_number_of_local_adj_matrix_partition_rows();
     local_num_cols_     = current_graph_view_.get_number_of_local_adj_matrix_partition_cols();
-    local_num_edges_    = new_src_v.size();
+    base_vertex_id_     = current_graph_view_.get_local_vertex_first();
+
+    local_num_edges_ = edge_t{0};
+    for (size_t i = 0; i < current_graph_view_.get_number_of_local_adj_matrix_partitions(); ++i) {
+      local_num_edges_ += current_graph_view_.get_number_of_local_adj_matrix_partition_edges(i);
+    }
+
+    src_indices_v_.resize(local_num_edges_);
 
     cugraph::detail::offsets_to_indices(
       current_graph_view_.offsets(), local_num_rows_, src_indices_v_.data().get());
-  }
-#endif
 
-  std::
-    tuple<rmm::device_vector<vertex_t>, rmm::device_vector<vertex_t>, rmm::device_vector<weight_t>>
-    combine_local_edges(rmm::device_vector<vertex_t> &src_v,
-                        rmm::device_vector<vertex_t> &dst_v,
-                        rmm::device_vector<weight_t> &weight_v)
-  {
-    thrust::stable_sort_by_key(
-      rmm::exec_policy(stream_)->on(stream_),
-      dst_v.begin(),
-      dst_v.end(),
-      thrust::make_zip_iterator(thrust::make_tuple(src_v.begin(), weight_v.begin())));
-    thrust::stable_sort_by_key(
-      rmm::exec_policy(stream_)->on(stream_),
-      src_v.begin(),
-      src_v.end(),
-      thrust::make_zip_iterator(thrust::make_tuple(dst_v.begin(), weight_v.begin())));
+    rmm::device_uvector<vertex_t> numbering_indices(numbering_map.size(), stream_);
+    thrust::sequence(rmm::exec_policy(stream_)->on(stream_),
```

Comment:
Why creating O(n) sequences rather than using O(1) `counting_iterator`? (above and in a few other places)

Reply:
`cugraph::experimental::relabel` expects `std::tuple<vertex_t const *, vertex_t const *>` as the second parameter (where this is used).  I don't know how to make a counting iterator that provides a `vertex_t const *`.

The original code (before converting to using `relabel`) did in place sorts so the sequence needed to be realized.  If there is a way to convert it to a counting iterator, I'd be happy to try it.

Reply:
True, `thrust::sort()`, `thrust::sort_by_key()` expect a modifiable range (hence cannot be counting iterators).